### PR TITLE
test(general): Add test for snapshot fix command

### PR DESCRIPTION
**Pull request by chaitalisg**
This PR adds a new test package to test `snapshot fix` command. 
The test 
1. creates a kopia repository in file system
2. generates random data to be used as source data for snapshots
3. deletes a random pack blob to simulate repository corruption
4. uses snapshot fix command to repair the repository
5. verifies that the repository is operational after the repair 

The orchestration script recovery-job.sh can be used to run the test in a pipeline.

Testing repository in object store is a future task.

### CODE REVIEW

File: tools/robustness-job.sh [hunk 1]
```diff
@@ -106,7 +106,11 @@ readonly test_flags="-v -timeout=${test_timeout}\
 
 # Set the make target based on ENGINE_MODE
 ENGINE_MODE="${ENGINE_MODE:-}"
-make_target="robustness-tests"
+if [[ "${ENGINE_MODE}" == "" ]]; then
+    make_target="recovery-tests"
+else
+    make_target="robustness-tests"
+fi
```
Comment (julio-lopez) [inline tools/robustness-job.sh:113]: I assume this outdated. Given the `tools/fault-tolerance-job.sh` script above, there's probably no need for this change here.

File: tools/fault-tolerance-job.sh [hunk 2]
```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env bash
```
Comment (julio-lopez) [inline tools/fault-tolerance-job.sh:1]: rename to `tools/recovery-fix-test-job.sh` or similar

File: tools/robustness-job.sh (hunk 1 above)
Comment (chaitalisg) [reply tools/robustness-job.sh:113]: That's the plan, yes. I am trying to make the pipeline work with the new job.sh. Once that is successful, I will commit the updates and remove the robustness-job.sh.

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
+package blobmanipulator
+
+// Package blobmanipulator provides the framework for snapshot fix testing
```
Comment (Shrekster) [inline tests/recovery/blobmanipulator/blobmanipulator.go:3]: This can go on the top of this file I think.

Comment (Shrekster) [inline tests/recovery/blobmanipulator/blobmanipulator.go:3]: This can go on the top of this file I think.

File: tests/recovery/recovery_test/recovery_test.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+			log.Println("Error creating Blob Manipulator:", err)
+		}
+	}
+
+	bm.DirCreater = getSnapshotter(baseDir)
+	bm.ConnectOrCreateRepo(dataRepoPath)
+	bm.DataRepoPath = dataRepoPath
+
+	// populate the kopia repo under test with random snapshots
+	bm.SetUpSystemUnderTest()
```
[the remark this answers is not in the document]
Comment (chaitalisg) [reply tests/recovery/recovery_test/recovery_test.go:40]: You mean in addition to `os.Exit(result)` in `main_test`? Let me look into this.

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+		return false
+	}
+
+	bm.fileWriter = fw
+
+	return true
+}
+
+// RestoreGivenOrRandomSnapshot restores a given or a random snapshot from kopia repository into the provided target directory.
+func (bm *BlobManipulator) RestoreGivenOrRandomSnapshot(snapID, restoreDir string) (string, error) {
```
Comment (adowair) [inline tests/recovery/blobmanipulator/blobmanipulator.go:122]: Do we plan to use `snapID` in a future iteration? It looks like in all instances this function is called with an empty argument.

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+			return "", err
+		}
+
+		// Select a random snapshot to restore
+		snapID = string(s[rand.Intn(len(s))].ID)
+	}
+
+	_, msg, err := bm.KopiaCommandRunner.Run("snapshot", "restore", snapID, restoreDir)
+	if err != nil {
+		return msg, err
```
Comment (adowair) [inline tests/recovery/blobmanipulator/blobmanipulator.go:148]: Would it be better to return the deleted snap ID from this function, rather of the full `stdout`? that way, the concern of parsing the output remains contained. Either way, can we please document the output name/usage?

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	// create random data
+	bm.getFileWriter()
+
+	fileSize := 100
+	numFiles := 100
+
+	fileWriteOpts := map[string]string{
+		fiofilewriter.MaxDirDepthField:         strconv.Itoa(1),
```
Comment (adowair) [inline tests/recovery/blobmanipulator/blobmanipulator.go:169]: Does this have an effect on the test? I'm curious if increasing the depth could cover more edge cases

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+	}, nil
+}
+
+// ConnectOrCreateRepo connects to an existing repository if possible or creates a new one.
+func (bm *BlobManipulator) ConnectOrCreateRepo(dataRepoPath string) error {
+	return bm.DirCreater.ConnectOrCreateRepo(dataRepoPath)
+}
+
+// DeleteBlob deletes the provided blob or a random blob, in kopia repo.
+func (bm *BlobManipulator) DeleteBlob(blobID string) error {
```
Comment (adowair) [inline tests/recovery/blobmanipulator/blobmanipulator.go:52]: Same here: looks like `blobID` is always provided as `""` in this PR...

File: tests/recovery/recovery_test/recovery_test.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+	// delete random blob
+	// assumption: the repo contains "p" blobs to delete, else the test is a no-op
```
Comment (adowair) [inline tests/recovery/recovery_test/recovery_test.go:43]: Is it ever possible for this to happen? If so, will the rest of the test run/fail as expected?

**Review by adowair — COMMENTED** [review]
Thanks @chaitalisg, some questions below

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+	_, err = bm.fileWriter.WriteRandomFiles(ctx, fileWriteOpts)
+	if err != nil {
+		return
+	}
+
+	// create snapshot of the data
+	snapPath := bm.fileWriter.DataDirectory(ctx)
+	log.Printf("Creating snapshot of directory %s", snapPath)
+
+	_, err = bm.TakeSnapshot(snapPath)
```
Comment (aaron-kasten) [inline tests/recovery/blobmanipulator/blobmanipulator.go:187]: Verify that snapshot succeeded:  either inspect or restore snapshot and verify that contents match results of `WriteRandomFiles`

File: tests/recovery/recovery_test/recovery_test.go [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+	baseDir := makeDir("base-dir-")
+
+	bm, err := blobmanipulator.NewBlobManipulator(baseDir)
+	if err != nil {
+		if errors.Is(err, kopiarunner.ErrExeVariableNotSet) {
+			log.Println("Skipping robustness tests because KOPIA_EXE is not set")
+		} else {
+			log.Println("Error creating Blob Manipulator:", err)
+		}
```
Comment (aaron-kasten) [inline tests/recovery/recovery_test/recovery_test.go:33]: Shouldn't you fail the test right about here?

File: tests/recovery/recovery_test/recovery_test.go [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+	bm.DirCreater = getSnapshotter(baseDir)
+	bm.ConnectOrCreateRepo(dataRepoPath)
+	bm.DataRepoPath = dataRepoPath
+
+	// populate the kopia repo under test with random snapshots
+	bm.SetUpSystemUnderTest()
+
+	// delete random blob
+	// assumption: the repo contains "p" blobs to delete, else the test is a no-op
+	bm.DeleteBlob("")
```
Comment (aaron-kasten) [inline tests/recovery/recovery_test/recovery_test.go:45]: Check `err` result of `DeleteBlob`.  Fail the test if it didn't succeed.

File: tests/recovery/recovery_test/recovery_test.go [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+	"github.com/kopia/kopia/tests/recovery/blobmanipulator"
+	"github.com/kopia/kopia/tests/robustness/snapmeta"
+	"github.com/kopia/kopia/tests/tools/kopiarunner"
+)
+
+func TestSnapshotFix(t *testing.T) {
+	// assumption: the test is run on filesystem and not directly on object store
+	dataRepoPath := path.Join(*repoPathPrefix, dataSubPath)
+
+	baseDir := makeDir("base-dir-")
```
Comment (aaron-kasten) [inline tests/recovery/recovery_test/recovery_test.go:25]: Check for failure of `makeDir`.  `makeDir` will return `""` on failure.  Abort test if `makeDir` fails.

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	return &BlobManipulator{
+		KopiaCommandRunner: runner,
+	}, nil
+}
+
+// ConnectOrCreateRepo connects to an existing repository if possible or creates a new one.
+func (bm *BlobManipulator) ConnectOrCreateRepo(dataRepoPath string) error {
+	return bm.DirCreater.ConnectOrCreateRepo(dataRepoPath)
```
Comment (aaron-kasten) [inline tests/recovery/blobmanipulator/blobmanipulator.go:48]: `DirCreater` may be `nil` if call that sets it fails.  You must check at the point that `DirCreater` is set for errors.  Do not continue to this line of code if `DirCreater` is `nil`.

**Review by aaron-kasten — CHANGES_REQUESTED** [review]
You must check for error results at every function call that involved File or Network I/O.  You must make a decision to either continue, or fail the test.  Normally, you would want to fail the test on I/O error.
Use the "rubber duck" method for inspecting your code.  Go through each line as if running the code (as if explaining what each line does to a rubber duck).  As yourself if the test should continue of fail based on the result of each call.  If the test should fail, then ensure there's a means of aborting the test - either `return err`, `exit(1)` or `t.Fail` depending on the context.

File: tests/recovery/blobmanipulator/blobmanipulator.go (hunk 14 above)
Comment (chaitalisg) [reply tests/recovery/blobmanipulator/blobmanipulator.go:48]: done

File: tests/recovery/recovery_test/recovery_test.go (hunk 11 above)
Comment (chaitalisg) [reply tests/recovery/recovery_test/recovery_test.go:33]: done

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+// RestoreGivenOrRandomSnapshot restores a given or a random snapshot from kopia repository into the provided target directory.
+func (bm *BlobManipulator) RestoreGivenOrRandomSnapshot(snapID, restoreDir string) (string, error) {
+	err := bm.KopiaCommandRunner.ConnectRepo("filesystem", "--path="+bm.DataRepoPath)
+	if err != nil {
+		return "", err
+	}
+
+	if snapID == "" {
+		// list available snaphsots
+		stdout, _, snapshotListErr := bm.KopiaCommandRunner.Run("snapshot", "list", "--json")
```
[the remark this answers is not in the document]
Comment (chaitalisg) [reply tests/recovery/blobmanipulator/blobmanipulator.go:131]: I tried to do this. The linter throws shadowing error if I use `err` instead of a new variable.

File: tests/recovery/blobmanipulator/blobmanipulator.go (hunk 7 above)
Comment (chaitalisg) [reply tests/recovery/blobmanipulator/blobmanipulator.go:170]: Yes. I will add more depth once this basic test is in.

File: tests/recovery/blobmanipulator/blobmanipulator.go (hunk 5 above)
Comment (chaitalisg) [reply tests/recovery/blobmanipulator/blobmanipulator.go:123]: Yes. IT's a provision to add more tests in future.

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+
+	return "", nil
+}
+
+// SetUpSystemUnderTest connects or creates a kopia repo, writes random data in source directory,
+// creates snapshots of the source directory.
+func (bm *BlobManipulator) SetUpSystemUnderTest() {
+	err := bm.ConnectOrCreateRepo(bm.DataRepoPath)
+	if err != nil {
+		return
```
[the remark this answers is not in the document]
Comment (chaitalisg) [reply tests/recovery/blobmanipulator/blobmanipulator.go]: No. Returned the error from the function.

File: tests/recovery/recovery_test/recovery_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (chaitalisg) [reply tests/recovery/recovery_test/recovery_test.go]: Handled the error scenario.

File: tests/recovery/blobmanipulator/blobmanipulator.go (hunk 6 above)
Comment (chaitalisg) [reply tests/recovery/blobmanipulator/blobmanipulator.go:158]: I had the same thought. However, the `snapshot fix` command does not accept the `snapID` as an argument. It needs the `object ID` mentioned in `stdout`.

File: tests/recovery/blobmanipulator/blobmanipulator.go (hunk 10 above)
Comment (chaitalisg) [reply tests/recovery/blobmanipulator/blobmanipulator.go:201]: I will do this in a subsequent PR.

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	// create random data
+	bm.getFileWriter()
+
+	fileSize := 100
+	numFiles := 100
+
+	fileWriteOpts := map[string]string{
```
[the remark this answers is not in the document]
Comment (chaitalisg) [reply tests/recovery/blobmanipulator/blobmanipulator.go:182]: Actually, I want the sequence to create dependent pack blobs. If I remove one of these, the repository is not big enough to simulate corruption.

**Review by aaron-kasten — APPROVED** [review]
Looks good.

File: tests/recovery/recovery_test/recovery_test.go (hunk 9 above)
Comment (chaitalisg) [reply tests/recovery/recovery_test/recovery_test.go]: Handled this in the code

File: tests/recovery/recovery_test/recovery_test.go [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+	if baseDir == "" {
+		t.FailNow()
+	}
+
+	bm, err := blobmanipulator.NewBlobManipulator(baseDir)
+	if err != nil {
+		if errors.Is(err, kopiarunner.ErrExeVariableNotSet) {
+			log.Println("Skipping robustness tests because KOPIA_EXE is not set")
```
Comment (redgoat650) [inline tests/recovery/recovery_test/recovery_test.go:33]: Does this need `t.Skip` instead? the log message seems to imply skip.
```suggestion
			t.Skip("Skipping robustness tests because KOPIA_EXE is not set")
```

File: tests/recovery/recovery_test/recovery_test.go [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+	"github.com/kopia/kopia/tests/recovery/blobmanipulator"
+	"github.com/kopia/kopia/tests/robustness/snapmeta"
+	"github.com/kopia/kopia/tests/tools/kopiarunner"
+)
+
+func TestSnapshotFix(t *testing.T) {
+	// assumption: the test is run on filesystem & not directly on object store
+	dataRepoPath := path.Join(*repoPathPrefix, dataSubPath)
+
+	baseDir := makeDir("base-dir-")
```
Comment (redgoat650) [inline tests/recovery/recovery_test/recovery_test.go:25]: Could you just call `t.TempDir()` instead? That automatically terminates the test on failure

File: tests/recovery/recovery_test/main_test.go [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+
+	// run the tests
+	result := m.Run()
+
+	os.Exit(result)
+}
+
+type kopiaRecoveryTestHarness struct {
+	dataRepoPath string
+	metaRepoPath string
```
Comment (redgoat650) [inline tests/recovery/recovery_test/main_test.go:35]: `metaRepoPath` doesn't seem to be used anywhere?

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 21]
```diff
@@ -0,0 +1,288 @@
...
+
+	return "", nil
+}
+
+// SnapshotFixRemoveFilesByFilename runs snapshot fix remove-files command with a provided file name.
+func (bm *BlobManipulator) SnapshotFixRemoveFilesByFilename(filename string) (string, error) {
+	// Get hold of the filename that can be used to in the snapshot fix command
+	stdout, msg, err := bm.KopiaCommandRunner.Run("snapshot", "fix", "remove-files", "--filename="+filename, "--commit")
+	if err != nil {
+		log.Println(stdout, msg)
```
Comment (redgoat650) [inline tests/recovery/blobmanipulator/blobmanipulator.go:283]: `Run` method I think already logs internally, if I'm not mistaken? Do we need to log again here? Same question for situations above.

File: tests/recovery/blobmanipulator/blobmanipulator.go [hunk 22]
```diff
@@ -0,0 +1,288 @@
...
+		fiofilewriter.MaxFileSizeField:         strconv.Itoa(fileSize),
+		fiofilewriter.MinFileSizeField:         strconv.Itoa(fileSize),
+		fiofilewriter.MaxNumFilesPerWriteField: strconv.Itoa(numFiles),
+		fiofilewriter.MinNumFilesPerWriteField: strconv.Itoa(numFiles),
+	}
+
+	_, err = bm.fileWriter.WriteRandomFiles(ctx, fileWriteOpts)
+	if err != nil {
+		return err
+	}
```
Comment (redgoat650) [inline tests/recovery/blobmanipulator/blobmanipulator.go:238]: Seems these repeated blocks of code could be decomposed into simpler methods to make `SetUpSystemUnderTest` more readable, e.g.
```go
err := bm.writeRandomFiles(fileSize, numFiles)
```

**Comment by Shrekster** [comment]
LGTM, thank you!
